doc/ir-semantics.text

```diff
@@ -53,63 +75,77 @@ succ I l =
 pred I l =
   { l' ∈ I : succ I l' = l }
 
-E ::= (x ↦ const v | x ↦ var a | x ↦ ⊥)*  lexical environment
-H ::= (l ↦ v)                             mutable heap
-
-T ::= (lit)*                    output trace
-
-Configuration: (T, H, E, P, I, l)
+Linear successor instruction
+  stepi (C* (s, l)) = (C* (succ P(s), l))
 
 Lookup (partial) function, returns a v:
   (H,E)[x] :=
-    v    if E ∋ (x ↦const v)
-    H(l) if E ∋ (x ↦mut l)
+    v    if E ∋ (x ↦ const v)
+    H(l) if E ∋ (x ↦ mut l)
 
 Update (partial) function, returns a S:
   (H,E)[x ← v] :=
-    H[E(x) ↦ v] if E ∋ (x ↦mut l)
+    H[E(x) ↦ v] if E ∋ (x ↦ mut l)
+
+## Evaluation of simple expressions:
+
+  eval P H E x = (H,E)[x]
+  eval P H E lit = lit
+  eval P H E primop(v, ...) = 〚primop〛(v, ...)
+
+## Evaluation of extended expressions
+
+  eeval P H E ⊥ = ⊥
+  eeval P H E e = eval P H E e
+
+## Reduction relation '-->'
 
-Evaluation of simple expressions:
+  (P, T, H, E*, C* C) --> M
+    when M := step P(C) (P, T, H, E*, C* C)
 
-  eval H E x = (H,E)[x]
-  eval H E lit = lit
-  eval H E primop(x₁, ..., xₙ) = 〚primop〛((H,E)[x₁], ..., (H,E)[xₙ])
+  step   CC[e]
+         (P, T, H, E*, C*)
+       = step CC[v] (P, T, H, E*, C*)
+    when v := eval P H E e
 
-Evaluation of extended expressions
+  step   (print x)
+         (P, T, H, E*, C*)
+       = (P, (T, (H,E)[x]), H, E*, stepi C*)
 
-  eeval H E ⊥ = ⊥
-  eeval H E e = eval H E e
+  step   (branch v l₁ l₂)
+         (P, T, H, E*, C* (s, l))
+       = (P, T, H, E*, C* (s, l₁))
+    when v = true
 
-Reduction relation
-  (T, H, E, P, I, l) -> ((T, (H,E)[x]), H, E, P, I, l + 1)
-    when I(l) = print x
+  step   (branch v l₁ l₂)
+         (P, T, H, E*, C* (s, l))
+       = (P, T, H, E*, C* (s, l₂))
+    when v = false
 
-  (T, H, E, P, I, l) -> (T, H, E, P, I, l')
-    when I(l) = branch x l₁ l₂
-     and l' := if (H,E)[x] then l₁ else l₂
+  step   (const x = v)
+         (P, T, H, E* E, C*)
+       = (P, T, H, E* (E, x ↦ v), stepi C*)
 
-  (T, H, E, P, I, l) -> (T, H, (E, x ↦ v), P, I, succ I l)
-    when I(l) = (const x = e)
-     and v := eval H E x
+  step   (mut x = v)
+         (P, T, H, E* E, C*)
+       = (P, T, H[a ↦ v], E* E[x ↦ v], stepi C*)
+    when a fresh
 
-  (T, H, E, P, I, l) -> (T, (H, a ↦ v), (E, x ↦ a), P, I, succ I l)
-    when I(l) = (mut x = e)
-     and v := eval H E e
-     and a fresh
+  step   (x ← v)
+         (P, T, H, E* E, C*)
+       = (P, T, (H,E)[x] := v, E* E, stepi C*)
```

Comment:
Looking at the definition on line 87, I guess the notation should be `(H,E)[x ← v]`